src/jetton.c

```diff
@@ -0,0 +1,367 @@
+                                 const address_t *master) {
+    BitString_t bits;
+    CellRef_t refs[2];
+    refs[0] = *code;
+
+    BitString_init(&bits);
+    BitString_storeCoins(&bits, 0);
+    BitString_storeAddress(&bits, owner->chain, owner->hash);
+    BitString_storeAddress(&bits, master->chain, master->hash);
+    hash_Cell(&bits, refs, 1, &refs[1]);
+
+    BitString_init(&bits);
+    BitString_storeUint(&bits, 0b00110, 5);
+    hash_Cell(&bits, refs, 2, out);
+}
+
+static bool jetton_state_assembler_dispatcher(CellRef_t *out,
+                                              const CellRef_t *code,
+                                              const address_t *owner,
+                                              const address_t *master,
+                                              uint8_t state_assembler_idx) {
+    switch (state_assembler_idx) {
+        case 0:
+            assemble_usdt_state(out, code, owner, master);
+            return true;
+        case 1:
+            assemble_tston_state(out, code, owner, master);
+            return true;
+        case 2:
+            assemble_wston_state(out, code, owner, master);
+            return true;
+        case 3:
+            assemble_hton_state(out, code, owner, master);
+            return true;
+        case 4:
+            assemble_stton_state(out, code, owner, master);
+            return true;
+    }
+    return false;
+}
+
+static bool jetton_get_id_by_name(const char *name, uint8_t *id) {
+    for (uint8_t idx = 0U; idx < ARRAY_SIZE(jettons); idx++) {
+        if (strncmp(jettons[idx].name, name, JETTON_NAME_MAX_SIZE) == 0) {
+            *id = idx;
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool jetton_get_wallet_address(size_t jetton_id, const address_t *owner, address_t *jetton_wallet) {
+    CellRef_t state_init;
+    address_t master;
+
+    if (jetton_id > ARRAY_SIZE(jettons)) {
+        return false;
+    }
+
+    if (owner == NULL) {
+        return false;
+    }
+
+    if (jetton_wallet == NULL) {
+        return false;
+    }
+
+    explicit_bzero(jetton_wallet, sizeof(address_t));
+
+    memcpy(master.hash, jettons[jetton_id].master_hash, HASH_LEN);
+    master.chain = jettons[jetton_id].master_workchain;
+
+    SAFE(jetton_state_assembler_dispatcher(&state_init,
+                                           &jettons[jetton_id].code,
+                                           owner,
+                                           &master,
+                                           jettons[jetton_id].state_assembler_idx));
+
+    jetton_wallet->chain = master.chain;
+    memcpy(jetton_wallet->hash, state_init.hash, HASH_LEN);
+
+    return true;
+}
+
+const char *jetton_get_name(size_t jetton_id) {
+    if (jetton_id > ARRAY_SIZE(jettons)) {
+        return "";
+    }
+
+    return jettons[jetton_id].name;
+}
+
+uint8_t jetton_get_decimals(size_t jetton_id) {
+    if (jetton_id > ARRAY_SIZE(jettons)) {
+        return 0U;
```

Comment:
Same, I think that if we want to handle the case where we don't know the jetton separatly